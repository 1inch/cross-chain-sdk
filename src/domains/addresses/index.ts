export {EvmAddress} from './evm-address.js'
export {SolanaAddress} from './solana-address.js'
export {AddressLike} from './types.js'
export {createAddress} from './address.factory.js'
