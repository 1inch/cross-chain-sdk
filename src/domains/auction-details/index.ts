export * from './auction-details'
export * from './types'
