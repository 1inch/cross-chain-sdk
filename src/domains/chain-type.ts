export enum ChainType {
    EVM,
    SVM
}
