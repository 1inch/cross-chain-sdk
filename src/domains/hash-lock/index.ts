export {HashLock, MerkleLeaf} from './hash-lock'
