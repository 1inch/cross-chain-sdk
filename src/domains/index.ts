export * from './addresses'
export * from './auction-details'
export * from './hash-lock'
export * from './time-locks'
export * from './chain-type'
export * from './immutables'
