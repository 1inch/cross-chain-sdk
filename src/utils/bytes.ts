import {isHexBytes} from '@1inch/byte-utils'
import assert from 'assert'

export function bufferFromHex(hex: string): Buffer {
    assert(isHexBytes(hex))

    return Buffer.from(hex.slice(2), 'hex')
}
