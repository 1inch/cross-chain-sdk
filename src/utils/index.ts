export * from './id'
export * from './validation'
export * from './time'
export * from './math'
export * from './addresses'
