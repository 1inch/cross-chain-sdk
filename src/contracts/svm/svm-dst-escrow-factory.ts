import {BN, BorshCoder, web3} from '@coral-xyz/anchor'
import assert from 'assert'
import {Buffer} from 'buffer'
import {Instruction} from './instruction'
import {BaseProgram} from './base-program'
import {WhitelistContract} from './whitelist'
import {ParsedCreateDstEscrowInstructionData} from './types'
import {bigintToBN} from '../../utils/numbers/bigint-to-bn'
import {uintAsBeBytes} from '../../utils/numbers/uint-as-be-bytes'
import {uint256split} from '../../utils/numbers/uint256-split'
import {getAta, getPda} from '../../utils'
import {
    EvmAddress,
    HashLock,
    Immutables,
    SolanaAddress,
    TimeLocks
} from '../../domains'
import {IDL} from '../../idl/cross-chain-escrow-dst'
import {FixedLengthArray} from '../../type-utils'
import {bufferToHex} from '../../utils/bytes'
import {bnArrayToBigInt} from '../../utils/numbers/bn-array-to-big-int'

export class SvmDstEscrowFactory extends BaseProgram {
    static DEFAULT = new SvmDstEscrowFactory(
        new SolanaAddress('GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve')
    )

    private static readonly coder = new BorshCoder(IDL)

    constructor(programId: SolanaAddress) {
        super(programId)
    }

    static parseCreateEscrowInstruction(
        ix: Instruction
    ): ParsedCreateDstEscrowInstructionData {
        const decoded = SvmDstEscrowFactory.coder.instruction.decode(
            ix.data
        ) as {
            name: string
            data: {
                orderHash: FixedLengthArray<number, 32>
                hashlock: FixedLengthArray<number, 32>
                amount: BN
                safetyDeposit: BN
                recipient: web3.PublicKey
                timelocks: FixedLengthArray<BN, 4>
                srcCancellationTimestamp: number
                assetIsNative: boolean
            }
        }

        assert(decoded, 'cannot decode create instruction')
        assert(decoded.name === 'create', 'not create instruction')

        const data = decoded.data

        return {
            orderHash: bufferToHex(data.orderHash),
            hashlock: HashLock.fromString(bufferToHex(data.hashlock)),
            amount: BigInt(data.amount.toString()),
            safetyDeposit: BigInt(data.safetyDeposit.toString()),
            recipient: EvmAddress.fromBuffer(data.recipient.toBuffer()),
            timelocks: TimeLocks.fromBigInt(bnArrayToBigInt(data.timelocks)),
            srcCancellationTimestamp: data.srcCancellationTimestamp,
            assetIsNative: data.assetIsNative
        }
    }

    static parsePrivateWithdrawInstruction(ix: Instruction): {secret: string} {
        const decoded = this.coder.instruction.decode(ix.data) as {
            name: string
            data: {secret: FixedLengthArray<number, 32>}
        }

        assert(decoded, 'cannot decode withdraw instruction')
        assert(decoded.name === 'withdraw', 'not withdraw instruction')

        return {
            secret: bufferToHex(decoded.data.secret)
        }
    }

    static parsePublicWithdrawInstruction(ix: Instruction): {secret: string} {
        const decoded = this.coder.instruction.decode(ix.data) as {
            name: string
            data: {secret: FixedLengthArray<number, 32>}
        }

        assert(decoded, 'cannot decode publicWithdraw instruction')
        assert(
            decoded.name === 'publicWithdraw',
            'not publicWithdraw instruction'
        )

        return {
            secret: bufferToHex(decoded.data.secret)
        }
    }

    public getEscrowAddress(params: EscrowAddressParams): SolanaAddress {
        return getPda(this.programId, [
            this.encoder.encode('escrow'),
            params.orderHash,
            params.hashLock.toBuffer(),
            params.maker.toBuffer(),
            uintAsBeBytes(params.amount, 64)
        ])
    }

    public createEscrow(
        params: Immutables<SolanaAddress>,
        extra: {
            /**
             * TokenProgram or TokenProgram 2022
             */
            tokenProgramId: SolanaAddress
            srcCancellationTimestamp: bigint
        }
    ): Instruction {
        const token = params.token.isNative()
            ? SolanaAddress.WRAPPED_NATIVE
            : params.token

        const data = SvmDstEscrowFactory.coder.instruction.encode('create', {
            orderHash: params.orderHash,
            hashlock: params.hashLock.toBuffer(),
            amount: new BN(params.amount.toString()),
            safetyDeposit: new BN(params.safetyDeposit.toString()),
            recipient: params.maker,
            timelocks: uint256split(params.timeLocks.build()).map(bigintToBN),
            srcCancellationTimestamp: Number(extra.srcCancellationTimestamp),
            assetIsNative: params.token.isNative()
        })

        const escrowAddress = this.getEscrowAddress(params)

        return new Instruction(
            this.programId,
            [
                {
                    // 1. taker
                    pubkey: params.taker,
                    isSigner: true,
                    isWritable: true
                },
                {
                    // 2. mint
                    pubkey: token,
                    isWritable: false,
                    isSigner: false
                },
                this.optionalAccount(
                    {
                        // 3. taker_ata
                        pubkey: getAta(
                            params.taker,
                            token,
                            extra.tokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    params.token.isNative()
                ),
                {
                    // 4. escrow
                    pubkey: escrowAddress,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 5. escrow ata
                    pubkey: getAta(escrowAddress, token, extra.tokenProgramId),
                    isWritable: true,
                    isSigner: false
                },

                {
                    // 6. associated_token_program
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 7. token_program
                    pubkey: extra.tokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 8. rent
                    pubkey: SolanaAddress.SYSVAR_RENT_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 9. system_program
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }

    public withdrawPrivate(
        params: Immutables<SolanaAddress>,
        secret: Buffer,
        extra: {
            /**
             * TokenProgram or TokenProgram 2022
             */
            tokenProgramId: SolanaAddress
        }
    ): Instruction {
        const token = params.token.isNative()
            ? SolanaAddress.WRAPPED_NATIVE
            : params.token
        const data = SvmDstEscrowFactory.coder.instruction.encode('withdraw', {
            secret
        })
        const escrow = this.getEscrowAddress(params)

        return new Instruction(
            this.programId,
            [
                {
                    // 1. taker
                    pubkey: params.taker,
                    isSigner: true,
                    isWritable: true
                },
                {
                    // 2. recipient
                    pubkey: params.maker,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 3. dst token
                    pubkey: token,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 4. escrow
                    pubkey: escrow,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 5. escrow_ata
                    pubkey: getAta(escrow, token, extra.tokenProgramId),
                    isSigner: false,
                    isWritable: true
                },
                this.optionalAccount(
                    {
                        // 6. recipient_ata
                        pubkey: getAta(
                            params.maker,
                            params.token,
                            extra.tokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    params.token.isNative()
                ),
                {
                    // 7. associated_token_program
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 8. token_program
                    pubkey: extra.tokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 9. system_program
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }

    public withdrawPublic(
        params: Immutables<SolanaAddress>,
        secret: Buffer,
        payer: SolanaAddress,
        extra: {
            /**
             * If not passed, than `WhitelistContract.DEFAULT` will be used
             * @see WhitelistContract.DEFAULT
             */
            whitelistProgramId?: SolanaAddress
            /**
             * TokenProgram or TokenProgram 2022
             */
            tokenProgramId: SolanaAddress
        }
    ): Instruction {
        const whitelistProgram = extra.whitelistProgramId
            ? new WhitelistContract(extra.whitelistProgramId)
            : WhitelistContract.DEFAULT

        const token = params.token.isNative()
            ? SolanaAddress.WRAPPED_NATIVE
            : params.token
        const data = SvmDstEscrowFactory.coder.instruction.encode(
            'publicWithdraw',
            {secret}
        )
        const escrow = this.getEscrowAddress(params)

        return new Instruction(
            this.programId,
            [
                {
                    // 1. creator
                    pubkey: params.taker,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 2. recipient
                    pubkey: params.maker,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 3. payer
                    pubkey: payer,
                    isSigner: true,
                    isWritable: true
                },
                {
                    // 4. resolver_access
                    pubkey: whitelistProgram.getAccessAccount(payer),
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 5. mint
                    pubkey: token,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 6. escrow
                    pubkey: escrow,
                    isSigner: false,
                    isWritable: true
                },
                {
                    // 7. escrow_ata
                    pubkey: getAta(escrow, token, extra.tokenProgramId),
                    isSigner: false,
                    isWritable: true
                },
                this.optionalAccount(
                    {
                        // 8. recipient_ata
                        pubkey: getAta(
                            params.maker,
                            params.token,
                            extra.tokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    params.token.isNative()
                ),
                {
                    // 9. associated_token_program
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 10. token_program
                    pubkey: extra.tokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                {
                    // 11. system_program
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }
}
export type EscrowAddressParams = Pick<
    Immutables<SolanaAddress>,
    'orderHash' | 'hashLock' | 'maker' | 'amount'
>
