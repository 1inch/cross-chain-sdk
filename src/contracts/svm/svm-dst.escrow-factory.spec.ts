import {now} from '@1inch/fusion-sdk'
import {SvmDstEscrowFactory} from './svm-dst-escrow-factory'
import {bufferFromHex} from '../../utils/bytes'
import {HashLock, Immutables, SolanaAddress, TimeLocks} from '../../domains'

describe('SVM Escrow dst factory', () => {
    it('should generate create escrow instruction', () => {
        const immutables = Immutables.new({
            orderHash: bufferFromHex(
                '0x663e58c85e0ac6721b6fe635eda03279e9552866f45938dc00effaac1fd0ac0b'
            ),
            hashLock: HashLock.fromString(
                '0x6e6b426ba6241e7544d5b6802897f93d68e6ed09cbc816058cf67096ff1494de'
            ),
            maker: SolanaAddress.fromBuffer(
                bufferFromHex(
                    '0xb96a60f72dcd59fa0e8378cd0f86bbda0a6dc305d1a228705c9d620ada73b62c'
                )
            ),
            taker: SolanaAddress.fromBuffer(
                bufferFromHex(
                    '0xaa298a629432ad57157beff906c3695e45871e077c33579c13db2d03f9607569'
                )
            ),
            token: SolanaAddress.fromBuffer(
                bufferFromHex(
                    '0x6521c75081a686617dfe3d3af3015864adc2860fc497d7dd470f522850c660a7'
                )
            ),
            amount: 1000000000n,
            safetyDeposit: 1000n,
            timeLocks:
                TimeLocks.fromBigInt(
                    47259175731581550628966830069363148924329979951954020279276990920496521412618n
                )
        })

        const ix = SvmDstEscrowFactory.DEFAULT.createEscrow(immutables, {
            srcCancellationTimestamp: now(),
            tokenProgramId: SolanaAddress.TOKEN_PROGRAM_ID
        })

        expect(ix).toMatchSnapshot()
    })
})
