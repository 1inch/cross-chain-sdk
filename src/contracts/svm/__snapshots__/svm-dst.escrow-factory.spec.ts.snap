// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SVM Escrow dst factory should generate create escrow instruction 1`] = `
Instruction {
  "accounts": [
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "CTF3k9CXyCjtCvpPMtG24Bo8r8qaQudTgqaSsp6A64tx",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "7on3yxZPgQgehEo4dxTNg2vV3ECWF8T8QVo3JJSGdcHk",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "6EdXXGGUP1nb4dzFkaTXXMAreSSDY18USHGBqT5raivs",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "EyYy71HpE9NhTczKyPX7TXCSKxwuhUkoF8B3LgEAphAo",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "AYMd52szbTMA7UwAkoT2yfp3WNbSeM5pUW6VhA4Y31sh",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "SysvarRent111111111111111111111111111111111",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": {
    "data": [
      24,
      30,
      200,
      40,
      5,
      28,
      7,
      119,
      102,
      62,
      88,
      200,
      94,
      10,
      198,
      114,
      27,
      111,
      230,
      53,
      237,
      160,
      50,
      121,
      233,
      85,
      40,
      102,
      244,
      89,
      56,
      220,
      0,
      239,
      250,
      172,
      31,
      208,
      172,
      11,
      110,
      107,
      66,
      107,
      166,
      36,
      30,
      117,
      68,
      213,
      182,
      128,
      40,
      151,
      249,
      61,
      104,
      230,
      237,
      9,
      203,
      200,
      22,
      5,
      140,
      246,
      112,
      150,
      255,
      20,
      148,
      222,
      0,
      202,
      154,
      59,
      0,
      0,
      0,
      0,
      232,
      3,
      0,
      0,
      0,
      0,
      0,
      0,
      185,
      106,
      96,
      247,
      45,
      205,
      89,
      250,
      14,
      131,
      120,
      205,
      15,
      134,
      187,
      218,
      10,
      109,
      195,
      5,
      209,
      162,
      40,
      112,
      92,
      157,
      98,
      10,
      218,
      115,
      182,
      44,
      10,
      0,
      0,
      0,
      210,
      0,
      0,
      0,
      54,
      1,
      0,
      0,
      154,
      1,
      0,
      0,
      10,
      0,
      0,
      0,
      110,
      0,
      0,
      0,
      210,
      0,
      0,
      0,
      221,
      190,
      123,
      104,
      124,
      245,
      125,
      104,
      0,
    ],
    "type": "Buffer",
  },
  "programId": "GveV3ToLhvRmeq1Fyg3BMkNetZuG9pZEp4uBGWLrTjve",
}
`;
